File: Dumplings/src/main/java/net/sinclairstudios/android/dumplings/domain/Fraction.java
package net.sinclairstudios.android.dumplings.domain;

public class Fraction implements Comparable<Fraction> {
    private final int numerator;
    private final int denominator;

    public Fraction(int numerator, int denominator) {
        this.numerator = numerator;
        this.denominator = denominator;
    }

    public int getAsInt() {
        return this.numerator / this.denominator;
    }

    private int getRemainderAsInt() {
        return this.numerator % this.denominator;
    }

    public Fraction getRemainder() {
        return new Fraction(getRemainderAsInt(), this.denominator);
    }

    public Fraction add(Fraction that) {
        assertSameDenominator(that);
        return new Fraction(this.numerator + that.numerator, this.denominator);
    }

    public Fraction subtract(Fraction that) {
        assertSameDenominator(that);
        return new Fraction(this.numerator - that.numerator, this.denominator);
    }

    public boolean hasRemainder() {
        return getRemainderAsInt() != 0;
    }

    @Override
    public int compareTo(Fraction that) {
        assertSameDenominator(that);
        return this.numerator - that.numerator;
    }

    private void assertSameDenominator(Fraction that) {
        if (that.denominator != this.denominator) {
            throw new IllegalArgumentException("fraction denominators must be the same");
        }
    }
}

File: Dumplings/src/main/kotlin/net/sinclairstudios/android/dumplings/domain/Fraction.kt
package net.sinclairstudios.android.dumplings.domain


public class Fraction(val numerator: Int, val denominator: Int) : Comparable<Fraction> {
    public fun getAsInt(): Int {
        return this.numerator / this.denominator
    }

    private fun getRemainderAsInt(): Int {
        return this.numerator % this.denominator
    }

    public fun getRemainder(): Fraction {
        return Fraction(getRemainderAsInt(), this.denominator)
    }

    public fun add(that: Fraction): Fraction {
        assertSameDenominator(that)
        return Fraction(this.numerator + that.numerator, this.denominator)
    }

    public fun subtract(that: Fraction): Fraction {
        assertSameDenominator(that)
        return Fraction(this.numerator - that.numerator, this.denominator)
    }

    public fun hasRemainder(): Boolean {
        return getRemainderAsInt() != 0
    }

    public override fun compareTo(other: Fraction): Int {
        assertSameDenominator(other)
        return this.numerator - other.numerator
    }

    private fun assertSameDenominator(that: Fraction): Unit {
        if (this.denominator != that.denominator) {
            throw IllegalArgumentException("fraction denominators must be the same")
        }
    }
}
